ads/google/utils.js

```diff
@@ -14,16 +14,289 @@
+    (adClient ? `&client=${encodeURIComponent(adClient.substr(0, 50))}` : '') +
+    (adSlot ? `&slotname=${encodeURIComponent(adSlot.substr(0, 50))}` : '') +
+    '&d_imp=1&output=html' +
+    `&format=${encodedWidth}x${encodedHeight}` +
+    `&w=${encodedWidth}&h=${encodedHeight}` +
+    `&adk=${adKey(slotNumber, slot, viewport, screen)}` +
+    // Last modified time.
+    // (lmt ? `&lmt=${(Date.parse(lmt) / 1000).toString()}` : '') +
+    // An identifier for the page view, used for offline logs analysis.
+    `&correlator=${correlator}` +
+    // Whether webGL is supported by the browser.
+    `&wgl=${global['WebGLRenderingContext'] ? '1' : '0'}` +
+    // The datetime of the request. Used to cache-proof URLs and for
+    // timing analysis.
+    `&dt=${startTime}` +
+    // Version of AMP runtime
+    '&amp_v=$internalRuntimeVersion$' +
+    // Durations used by CSI
+    // `&bpp=29&bdt=144&fdt=33&idt=42` +
+    // User time zone offset in minutes.
+    `&u_tz=${-(new Date).getTimezoneOffset()}` +
+    (screen ? `&u_w=${screen.width}&u_h=${screen.height}` +
+              `&u_aw=${screen.availWidth}&u_ah=${screen.availHeight}` +
+              `&u_cd=${screen.colorDepth}` : '') +
+    // Font and size. These require (forbidden) getComputedStyle.
+    // `&dff=${}&dfs=16` +
+    // Ad slot's x and y coordinates. (slot is relative to viewport.)
+    (viewport ? `&adx=${slot.x + viewport.x}&ady=${slot.y + viewport.y}` +
+                // Viewport width and height.
+                `&biw=${viewport.width}&bih=${viewport.height}` :
+                `&adx=${SIZE_EXCEPTION_ID}&ady=${SIZE_EXCEPTION_ID}` +
+                `&biw=${SIZE_EXCEPTION_ID}&bih=${SIZE_EXCEPTION_ID}`) +
+    // Have to think about how to handle publisher window.
+    // `&isw=300&ish=200` +
+    // Requires looking outside in the DOM.
+    // `&ifk=3286625729` +
+    // Resend count.
+    // `&rx=0` + `
+    // Visibility state.
+    `&vis=${visNum}` +
+    // Looks at parents, needs getComputedStyle.
+    // `&pfx=0` +
+    // Browser supports SVG?
+    (global.SVGElement && global.document.createElementNS ? '&bc=1' : '') +
+    // Unique ID within window.
+    `&ifi=${slotNumber}` +
+    // Experiment IDs, if any.
+    (experimentIds ? `&eid=${experimentIds}` : '');
+
+  let truncated = false;
+  const addParameter = (name, value) => {
+    if (value != undefined) {
+      const newParam = `&${name}=${encodeURIComponent(value)}`;
+      if (url.length + newParam.length <= MAX_URL_LENGTH_) {
+        url += newParam;
+      } else {
+        truncated = true;
+      }
+    }
+  };
+
+  const addTruncatableParameter = (name, value, minUsefulLength) => {
+    if (value != undefined) {
+      const newParam = `&${name}=${encodeURIComponent(value)}`;
+      const urlLength = url.length + newParam.length;
+      if (urlLength <= MAX_URL_LENGTH_) {
+        url += newParam;
+      } else {
+        // Truncate to fit.
+        url = url.substr(
```

Comment:
We're truncating the already built url, not the `newParam`. Is that intentional?